dns/dns.go

```diff
@@ -214,3 +221,216 @@ func serveUDPAPI(store *dnsStore) {
+	if err != nil {
+		log.Fatal("Update config unmarshal:", err)
+	}
+	// fmt.Println(newCluster.ClusterToken)
+	return append(clusters, newCluster)
+}
+
+// print out []jsonClusterInfo
+func PrintClusterConfig(clusters []jsonClusterInfo) {
+	for _, cluster := range clusters {
+		fmt.Println(cluster.ClusterToken)
+		for _, member := range cluster.Members {
+			fmt.Println("\t", member.NsName)
+			fmt.Println("\t", member.GlueRecord)
+		}
+	}
+}
+
+// send the updated cluster info to other Raft clusters
+// the last entry in clusters is the new cluster
+func sendClusterInfo(clusters []jsonClusterInfo, destIP string) {
+	// pick the first member ot send the cluster config to
+	addr := "http://" + destIP + ":9121/addcluster"
+	clusterJSON, err := json.Marshal(clusters)
+	log.Println("Sending cluster info to", addr)
+	if err != nil {
+		log.Fatal("sendClusterInfo: ", err)
+	}
+	req, err := http.NewRequest("PUT", addr, strings.NewReader(string(clusterJSON)))
+	if err != nil {
+		log.Fatal("sendClusterInfo: ", err)
+	}
+	req.ContentLength = int64(len(string(clusterJSON)))
+	resp, err := http.DefaultClient.Do(req)
+	log.Println("Received cluster info response", destIP)
+	if err != nil {
+		log.Fatal("sendClusterInfo: ", err)
+	}
+	if resp.StatusCode != http.StatusNoContent {
+		log.Fatal("sendClusterInfo request failed at server side")
+	}
+}
+
+func retrieveRR(clusters []jsonClusterInfo, store *dnsStore) {
+	var wg sync.WaitGroup
+	// excluding our own cluster
+	errorC := make(chan error)
+	for _, cluster := range clusters[:len(clusters)-1] {
+		member := cluster.Members[0]
+		t := strings.Split(member.GlueRecord, " ")
+		memberIP := t[len(t)-1]
+		addr := "http://" + memberIP + ":9121/getrecord"
+		wg.Add(1)
+		go func(addr string, store *dnsStore, errorC chan<- error, wg *sync.WaitGroup) {
+			defer wg.Done()
+			log.Println("Start retrieving RR from", addr)
+			counter := 0
+			for {
+				req, err := http.NewRequest("GET", addr, strings.NewReader(strconv.Itoa(counter)))
+				if err != nil {
+					errorC <- err
+					return
+				}
+				req.ContentLength = int64(len(strconv.Itoa(counter)))
+				resp, err := http.DefaultClient.Do(req)
+				if err != nil {
+					errorC <- err
+					return
+				}
+
+				body, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					errorC <- err
+					return
+				}
+				// parse body
+				var records []string
+				err = json.Unmarshal(body, &records)
+				if err != nil {
+					errorC <- err
+					return
+				}
+				log.Println("Received", records)
+				if len(records) == 1 && records[0] == "Done" {
+					// finished reading
+					log.Println("Finished reading RRs")
+					return
+				}
+				// go through Raft to add RRs
+				for _, rrString := range records {
+					if !checkValidRRString(rrString) {
+						// log.Fatal("Received bad RR")
+						errorC <- errors.New("Recevied bad RR")
+						return
+					}
+					// this is async, might cause too much traffic
+					store.ProposeAddRR(rrString)
```

Comment:
Add a small comment to explain why locks are not needed here (due to proposals done through the channel)